File: review.py
#!/usr/bin/env python

import os
import argparse
from git import Repo
from utils import get_repo, get_or_create_remote, validate_author_branch


def review(author, branch):
    repo = get_repo()

    if repo.is_dirty():
        print('The current branch has unstash changes')
        return

    remote = get_or_create_remote(repo, author)

    print('Fetching for new changes')
    remote.fetch(branch)
    
    print('Checkout to %s' % branch)
    if not branch in repo.branches:
        repo.git.checkout('-t', '%s/%s' % (author, branch))
    else:
        repo.git.checkout(branch)
        repo.git.reset('--hard', '%s/%s' % (author, branch))


def run():
    parser = argparse.ArgumentParser(description='Process some integers.')
    parser.add_argument('author_branch',
                        type=validate_author_branch,
                        help='<github_author>:<branch_name>')
    args = parser.parse_args()
    review(args.author_branch['author'], args.author_branch['branch'])


run()

File: src/git/review.ts
import { SimpleGit } from 'simple-git';
import { withArgs, warn, info, hasUnstashedChanges, addRemote } from './utils';

const ARG_PATTERN = /(?<remote>.+)\:(?<branch>.+)?/i;

async function run() {
  withArgs(ARG_PATTERN, async (git: SimpleGit, { remote, branch }) => {
    if (!remote || !branch) return;

    if (await hasUnstashedChanges(git)) {
      return warn('Please stash your current changes');
    }

    await addRemote(git, remote);

    info(`Fetching changes from ${remote}/${branch}`);
    await git.fetch(remote, branch);

    if (await doesBranchExists(git, branch)) {
      warn('Reseting branch ...');
      await git.checkout(branch);
      await git.reset(['--hard', `${remote}/${branch}`]);
    } else {
      info(`Checking out branch: ${branch}`);
      await git.checkout(['-t', `${remote}/${branch}`]);
    }
  });
};

async function doesBranchExists(git, branch) {
  const branches = (await git.branchLocal()).all;
  return branches.some(_branch => _branch === branch);
}

run();
